Add initial support for MICFIL PDM DAI

dbaluta opened the pull request:
This adds initial support for PDM MICFIL DAI found on NXP i.MX8MP.

dbaluta committed:
drivers: dai: Add new enum for PDM MICFIL type

Add new DAI type enum for PDM MICFIL.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

dbaluta committed:
drivers: dai: Add initial support for NXP MICFIL PDM IP

Introduce new DAI driver used for NXP's PDM MICFIL IP.
This block implements required digital interface to provide
a 24-bits audio signal from a PDM microphone bitstream in a configurable
output sampling rate.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

dbaluta committed:
dts: xtensa: nxp_imx8m: Add PDM MICFIL node

This adds micfil node for NPX i.MX8MP SOC.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

LaurentiuM1234 commented on drivers/dai/nxp/micfil/micfil.c:
```diff
@@ -0,0 +1,202 @@
+	((DT_INST_DMAS_CELL_BY_NAME(inst, dir, channel) & GENMASK(7, 0)) | \
+	((DT_INST_DMAS_CELL_BY_NAME(inst, dir, mux) << 8) & GENMASK(15, 8)))
+
+#define MICFIL_FIFO_DEPTH(inst) DT_INST_PROP(inst, fifo_depth)
```
don't think this is needed?

dbaluta replied:
do you mean should I directly use `DT_INST_PROP(inst, fifo_depth)` instead in the code?

LaurentiuM1234 replied:
yeah

LaurentiuM1234 commented on drivers/dai/nxp/micfil/micfil.c:
```diff
@@ -0,0 +1,202 @@
+{
+	struct dai_nxp_micfil_data *micfil_data = dev->data;
+
+	PDM_Reset(micfil_data->base);
```
is it a good idea to restart the module on TRIG_START? what if we PAUSE-RESUME? won't this lead to data loss?

dbaluta replied:
Yes, I think you are right. This pattern (of resetting the device at start) is followed by the Linux kernel also but they save the registers at suspend and restore them at resume.

So, I think I will reset the device at init() once only. Also, we use the software reset happens in PDM_Init() which is called by set_config() so I think it is not even necessary in start().

LaurentiuM1234 replied:
pls drop if not needed. Otherwise, I'd suggest doing it in `config()` unless there's some other reason we'd need to ungate clocks in init().

LaurentiuM1234 commented on drivers/dai/nxp/micfil/micfil.c:
```diff
@@ -0,0 +1,202 @@
+				  enum dai_trigger_cmd cmd)
+{
+	if (dir != DAI_DIR_RX) {
+		LOG_ERR("%s: invalid direction %d", __func__, dir);
```
same

LaurentiuM1234 commented on drivers/dai/nxp/micfil/micfil.c:
```diff
@@ -0,0 +1,202 @@
+	case DAI_TRIGGER_COPY:
+		return 0;
+	default:
+		LOG_ERR("%s: invalid trigger cmd %d", __func__, cmd);
```
same

LaurentiuM1234 replied:
can drop function name from the log

LaurentiuM1234 commented on drivers/dai/nxp/micfil/micfil.c:
```diff
@@ -0,0 +1,202 @@
+	.base = UINT_TO_MICFIL(DT_INST_REG_ADDR(inst)),				\
+	.cfg.type = DAI_IMX_MICFIL,						\
+	.cfg.dai_index = DT_INST_PROP_OR(inst, dai_index, 0),			\
+};										\
```
empty space here?

LaurentiuM1234 commented on drivers/dai/nxp/micfil/micfil.c:
```diff
@@ -0,0 +1,202 @@
+		chan_config.gain = kPDM_DfOutputGain2;
+		chan_config.cutOffFreq = kPDM_DcRemoverBypass;
+		PDM_SetChannelConfig(micfil_data->base, i, &chan_config);
+	}
```
empty space here?

LaurentiuM1234 commented on drivers/dai/nxp/micfil/micfil.c:
```diff
@@ -0,0 +1,202 @@
+	ret = PDM_SetSampleRateConfig(micfil_data->base, MICFIL_CLK_ROOT, bespoke->pdm_rate);
+	if (ret != kStatus_Success) {
+		LOG_ERR("Failure to set samplerate config rate %d", bespoke->pdm_rate);
+		return ret;
```
doesn't this need conversion to errno-like code? is ret < 0?

dbaluta replied:
True, this function can only return kStatus_Success or KStatus_Fail. Will return `-EINVAL` in case of failure.

LaurentiuM1234 reviewed (commented):
looks good, thanks!. Some nitpicks/questions below

dbaluta commented:
-- Changes in v2, addressed comments from @LaurentiuM1234

LaurentiuM1234 commented on drivers/dai/nxp/micfil/micfil.c:
```diff
@@ -0,0 +1,201 @@
+
+struct dai_nxp_micfil_config {
+	const struct dai_properties *rx_props;
+	uint32_t fifo_depth;
```
is this really needed? can't you just use `fifo_depth` from `rx_props`?

dbaluta replied:
dropped

LaurentiuM1234 commented on drivers/dai/nxp/micfil/micfil.c:
```diff
@@ -0,0 +1,201 @@
+	((DT_INST_DMAS_CELL_BY_IDX(inst, 0, mux) << 8) & GENMASK(15, 8)))
+
+struct dai_nxp_micfil_data {
+	PDM_Type *base;
```
can move this to `struct dai_nxp_micfil_config` unless this can be modified at runtime?

dbaluta replied:
will moved in v3.

LaurentiuM1234 reviewed (commented):
some more nits/questions below

dbaluta commented:
-- Changes since v3
* added support for configuring pinctrl from micfil zephyr driver, this will allow us to drop pinctrl from linux side.
*  dropped fifo_depth and used the one in rx_props
* dropped function name
* moved base to dai_nxp_micfil_config

LaurentiuM1234 reviewed (approved):
good work, thank you!
